common/djangoapps/student/models.py

```diff
@@ -2154,3 +2154,47 @@ def refund_window(self):
     def refund_window(self, refund_window):
         """Set the current refund window to the given timedelta."""
         self.refund_window_microseconds = int(refund_window.total_seconds() * 1000000)
+
+
+class UserAttribute(models.Model):
+    """
+    Record additional metadata about a user, stored as key/value pairs of text.
+    """
+
+    class Meta(object):
+        # Ensure that at most one value exists for a given user/name.
+        unique_together = (('user', 'name'))
+
+    user = models.ForeignKey(User, related_name='attributes')
+    name = models.CharField(max_length=255, help_text=_("Name of this user attribute."))
+    value = models.CharField(max_length=255, help_text=_("Value of this user attribute."))
+    created = models.DateTimeField(default=timezone.now)
+
+    def __unicode__(self):
+        """Unicode representation of this attribute. """
+        return u"[{username}] {name}: {value}".format(
+            name=self.name,
+            value=self.value,
+            username=self.user.username,
+        )
+
+    @classmethod
+    def set_user_attribute(cls, user, name, value):
+        """
+        Add an name/value pair as an attribute for the given
+        user. Overwrites any previous value for that name, if it
+        exists.
+        """
+        cls.objects.filter(user=user, name=name).delete()
```

Comment:
Fun note: `update_or_create` doesn't actually work with a `unique_together` constraint (possibly only on MySQL, but it's unclear to me yet). See https://build.testeng.edx.org/job/edx-platform-python-unittests-pr/18628/testReport/junit/student.tests.tests/UserAttributeTests/test_get_set_attribute_2/